services/galley/src/Galley/API/Teams.hs

```diff
@@ -856,141 +861,207 @@ canUserJoinTeam tid = do
+  ) =>
+  (TeamId -> Galley (Public.TeamFeatureStatus a)) ->
+  (TeamId -> Public.TeamFeatureStatus a -> Galley ()) ->
+  FeatureStatusHandlers (a :: Public.TeamFeatureName)
+mkFeatureStatusHandlers getter' setter' =
+  FeatureStatusHandlers
+    (mkGetFeatureStatusH getter')
+    (mkGetFeatureStatusInternalH getter')
+    (mkSetFeatureStatusH setter')
+    (mkSetFeatureStatusInternalH setter')
+  where
+    mkGetFeatureStatusH ::
+      (TeamId -> Galley (Public.TeamFeatureStatus a)) ->
+      UserId ::: TeamId ::: JSON ->
+      Galley Response
+    mkGetFeatureStatusH getter (uid ::: tid ::: _) = do
+      zusrMembership <- Data.teamMember tid uid
+      void $ permissionCheck (ViewTeamFeature (Public.knownTeamFeatureName @a)) zusrMembership
+      json <$> getter tid
+
+    mkGetFeatureStatusInternalH ::
+      (TeamId -> Galley (Public.TeamFeatureStatus a)) ->
+      TeamId ::: JSON ->
+      Galley Response
+    mkGetFeatureStatusInternalH getter (tid ::: _) = do
+      json <$> getter tid
+
+    mkSetFeatureStatusH ::
+      (TeamId -> Public.TeamFeatureStatus a -> Galley ()) ->
+      UserId ::: TeamId ::: JsonRequest (Public.TeamFeatureStatus a) ::: JSON ->
+      Galley Response
+    mkSetFeatureStatusH setter (uid ::: tid ::: req ::: _) = do
+      zusrMembership <- Data.teamMember tid uid
+      void $ permissionCheck (ChangeTeamFeature (Public.knownTeamFeatureName @a)) zusrMembership
+      status <- fromJsonBody req
+      setter tid status
+      pure (empty & setStatus status204)
+
+    mkSetFeatureStatusInternalH ::
+      (TeamId -> Public.TeamFeatureStatus a -> Galley ()) ->
+      TeamId ::: JsonRequest (Public.TeamFeatureStatus a) ::: JSON ->
+      Galley Response
+    mkSetFeatureStatusInternalH setter (tid ::: req ::: _) = do
+      status <- fromJsonBody req
+      setter tid status
+      pure (empty & setStatus status204)
+
+getSSOStatusInternal :: TeamId -> Galley (Public.TeamFeatureStatus 'Public.TeamFeatureSSO)
 getSSOStatusInternal tid = do
-  defConfig <- do
+  defStatus <- do
     featureSSO <- view (options . optSettings . setFeatureFlags . flagSSO)
     pure $ case featureSSO of
-      FeatureSSOEnabledByDefault -> Public.TeamFeatureEnabled
-      FeatureSSODisabledByDefault -> Public.TeamFeatureDisabled
-  ssoTeamConfig <- TeamFeatures.getFlag tid Public.TeamFeatureSSO
-  pure . Public.TeamFeatureStatus . fromMaybe defConfig $ ssoTeamConfig
-
-setSSOStatusInternal :: TeamId -> Public.TeamFeatureStatus -> Galley ()
-setSSOStatusInternal tid (Public.TeamFeatureStatus status) = do
-  case status of
+      FeatureSSOEnabledByDefault -> Public.mkFeatureStatus Public.TeamFeatureEnabled
+      FeatureSSODisabledByDefault -> Public.mkFeatureStatus Public.TeamFeatureDisabled
+  status <- TeamFeatures.getFeatureStatus @'Public.TeamFeatureSSO tid
+  pure . fromMaybe defStatus $ status
+
+setSSOStatusInternal :: TeamId -> (Public.TeamFeatureStatus 'Public.TeamFeatureSSO) -> Galley ()
+setSSOStatusInternal tid status = do
```

Comment:
I think we can write a function here that works for all `HasNoConfig` instances.

Reply:
Decided not to do because there's too much custom stuff each `*Internal` function does